geospace-lib/src/main/scala/com/socrata/geospace/lib/shapefile/LayerTransformer.scala

```diff
@@ -0,0 +1,117 @@
+ */
+case class LayerTransformer(projection: Projection) extends Logging{
+  type IngestResultMap = (String, (Traversable[Feature], Schema))
+
+  /**
+   * From an array of files, looks for a file with a given extension and returns success if found.
+   * @param directory Array to look for files with the given extension
+   * @param extension The desired file extension
+   * @return The first file in the array that matches the specified extension, or None if there are no matches.
+   */
+  private def getFileFromArray(directory: Array[File], extension: String): Either[InvalidShapefileSet, File] = {
+    // find file given the extension, if fond return that value, if not then send an InvalidShapeFIleSet
+    directory.find { f =>
+      FilenameUtils.getExtension(f.getName).equals(extension)
+    }.toRight(InvalidShapefileSet(s".$extension file is missing"))
+  }
+
+  /**
+   * Extracts the contents of a shapefile.
+   * Assumes that validate() has already been called on the shapefile contents.
+   * @param layerName name of layer containing the set of files that make up the shapefile
+   * @return The shapefile features and schema, reprojected to WGS84.
+   */
+  def transform(layerName: String, contents: Array[File]):
+                                                Either[InvalidShapefileSet, (Traversable[Feature], Schema)] = {
+    logMemoryUsage("Before reading Shapefile...")
+    // take each item, then push to transform.
+    parseShape(layerName, contents) match {
+      case Right(x) =>
+        Right(x)
+      case Left(e: InvalidShapefileSet) =>
+        Left(e)
+    }
+  }
+
+  /**
+   * Actual parsing of shapefiles done here. Including projections
+   */
+  private def parseShape(name: String, array: Array[File]):
+                                                  Either[InvalidShapefileSet,(Traversable[Feature], Schema)] = {
+    val contents = for {
+      shp  <- getFileFromArray(array, ShapeFileConstants.ShapeFormat).fold(Failure(_), Success(_))
+      proj <- Try(doProjections(projection, shp))
+    } yield proj
+
+    contents match {
+      case Success(c) =>
+        Right(c)
+      case Failure(e: Exception) =>
+        logger.warn("\"Reader failed to parse shape layer {}. -> {}", name, e.getMessage)
+        Left(InvalidShapefileSet("Reader failed to parse shape layer '%s'. -> %s".format(name, e.getMessage)))
+      case Failure(e) =>
+        throw e
+    }
+  }
+
+  /**
+   * performs feature and schema reprojections will not handle exceptions.
+   */
+  private def doProjections(projection: Projection, file: File): (Traversable[Feature], Schema) = {
+    val shapeFile = Shapefile(file)
+    try{
+      logger.info("Reprojecting shapefile schema and {} features to {}",
+        shapeFile.features.size.toString,
+        projection.getName)
+      logMemoryUsage("Before reprojecting features...")
+
+      var i = 0
+      // projecting features
+      val features: Traversable[Feature] = shapeFile.features.map { feature =>
```

Comment:
I wonder if we can make this an `Iterable` or `Iterator`.  Make it lazy so it doesn't take up all memory for all reprojected shapes.   I know this might be beyond your scope, but we should try to do it at some point to optimize memory usage.